File: code/cpp/ch03/solutions/warmup_05_sol.cpp
/*
 * Solution for Warmup 05: Sum 1 to N
 * ====================================
 * Chapter 3: Decisions and Loops
 *
 * This is the reference solution. Try to solve the problem yourself before
 * looking at this!
 *
 * APPROACH
 * --------
 * Use a for loop to accumulate the sum from 1 to n.
 * (There's a formula n*(n+1)/2, but the point here is to practice loops.)
 *
 * TIME COMPLEXITY:  O(n) — loop runs n times
 * SPACE COMPLEXITY: O(1) — just one accumulator variable
 */

#include <iostream>
using namespace std;

int solve(int n) {
    int sum = 0;
    for (int i = 1; i <= n; i++) {
        sum += i;
    }
    return sum;
}

// ── Do not change anything below this line ──────────────────────────
int main() {
    int n;
    cin >> n;
    cout << solve(n) << endl;
    return 0;
}

File: code/python/ch03/solutions/warmup_05_sol.py
"""
Solution for Warmup 05: Sum 1 to N
============================================
Chapter 3: Decisions and Loops

This is the reference solution. Try to solve the problem yourself before
looking at this!

APPROACH
--------
Use a loop to accumulate the sum from 1 to n. The formula n*(n+1)//2
would be O(1), but the point of this exercise is to practice loops.

TIME COMPLEXITY:  O(n) — one loop iteration per number
SPACE COMPLEXITY: O(1) — just the accumulator variable
"""


def solve(n: int) -> int:
    """Return the sum 1 + 2 + ... + n using a loop."""
    total = 0
    for i in range(1, n + 1):
        total += i
    return total


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    n = int(input())
    print(solve(n))
